src/Lean/Compiler/LCNF/FloatLetIn.lean

```diff
@@ -0,0 +1,263 @@
+/-
+Copyright (c) 2022 Henrik Böving. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Lean.Compiler.LCNF.CompilerM
+import Lean.Compiler.LCNF.FVarUtil
+import Lean.Compiler.LCNF.PassManager
+import Lean.Compiler.LCNF.Types
+
+namespace Lean.Compiler.LCNF
+
+namespace FloatLetIn
+
+/--
+The decision of the float mechanism.
+-/
+inductive Decision where
+|
+  /--
+  Push into the arm with name `name`.
+  -/
+  arm (name : Name)
+| /--
+  Push into the default arm.
+  -/
+  default
+|
+  /--
+  Dont move this declaration it is needed where it is right now.
+  -/
+  dont
+|
+  /--
+  No decision has been made yet.
+  -/
+  unknown
+deriving Hashable, BEq, Inhabited, Repr
+
+def Decision.ofAlt : Alt → Decision
+| .alt name _ _ => .arm name
+| .default _ => .default
+
+/--
+The context for `BaseFloatM`.
+-/
+structure BaseFloatContext where
+  /--
+  All the declarations that were collected in the current LCNF basic
+  block up to the current statement.
+  -/
+  decls : Array CodeDecl := #[]
```

Comment:
I think it is better using `List CodeDecl` or `PArray CodeDecl`. Types used with `ReaderT` are seldom used linearly. For example,  the function `withNewCandidate` will force the array to be copied since it is not an unique reference. If we had just a few calls to `withNewCandidate`, it would be fine to use `Array CodeDecl`, but this is not the case here.